docs/adr/ADR-012-typed-models-over-dicts.md

```diff
@@ -0,0 +1,545 @@
+    workspaces: list[Workspace]
+    dry_run: bool = True
+
+# Usage - full type safety and IDE support!
+request = ReconcileRequest(
+    workspaces=[
+        Workspace(
+            name="app-sre",
+            vault_token_path="slack/app-sre/token",
+            usergroups=[
+                Usergroup(
+                    handle="on-call",
+                    config=UsergroupConfig(
+                        users=["user1", "user2"],
+                        channels=["#alerts"],
+                    )
+                )
+            ]
+        )
+    ]
+)
+
+# IDE knows exact structure!
+workspace = request.workspaces[0]
+workspace.name  # ← IDE autocomplete works!
+workspace.usergroups[0].config.users  # ← Fully typed!
+```
+
+**Pros:**
+
+- **Full type safety:** mypy validates entire hierarchy
+- **Automatic validation:** Pydantic checks types and structure
+- **IDE autocompletion:** Works at every level
+- **Self-documenting:** Model hierarchy shows structure
+- **Easy testing:** Create test objects with known structure
+- **Refactoring safety:** Type changes propagate
+- **Validation features:** Can add validators, computed fields
+- **JSON serialization:** `.model_dump()` and `.model_validate()` built-in
+
+**Cons:**
+
+- More boilerplate (extra model classes)
+  - **Mitigation:** Boilerplate is documentation and type safety
+  - **Mitigation:** Models are reusable across codebase
+
+## Consequences
+
+### Positive
+
+- **Type safety:** mypy catches errors at development time, not runtime
+- **Automatic validation:** Pydantic validates structure and types on instantiation
+- **Immutability:** `frozen=True` prevents accidental modifications
+- **Thread-safe:** Frozen models are inherently thread-safe
+- **Self-documenting:** Model hierarchy is the documentation
+- **IDE support:** Full autocompletion and type hints at every level
+- **Easier testing:** Clear, testable object structure
+- **Refactoring safety:** Changes propagate through type system
+- **Runtime guarantees:** If object exists, structure is valid and unchanged
+- **Better error messages:** Pydantic shows exactly what's wrong
+- **Prevents unintended mutations:** No hidden state changes in deep call stacks
+
+### Negative
+
+- **More boilerplate:** Need to define model classes for each level
+  - **Mitigation:** Models serve as documentation
+  - **Mitigation:** Type safety worth the extra code
+  - **Mitigation:** Models are reusable
+
+- **Learning curve:** Developers must understand Pydantic
+  - **Mitigation:** Pydantic is standard in FastAPI ecosystem
+  - **Mitigation:** Document common patterns with examples
+
+- **Verbose for simple structures:** Overkill for very simple data
+  - **Mitigation:** Use TypedDict for truly simple cases (internal only)
+  - **Mitigation:** API contracts benefit from formality
+
+- **Cannot modify frozen models:** Need to use `.model_copy()` for changes
+  - **Mitigation:** Immutability prevents bugs - this is a feature, not a bug
```

Comment:
:heart: 

Im not sure what `mitigation` means in this or the other contexts in this section. To me this reads more like an explanation of why we choose to do this, rather than how we can work around it. More like a `reason`.